extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentPusher.java

```diff
@@ -99,13 +103,17 @@ public DataSegment push(File inDir, DataSegment segment) throws IOException
       size = CompressionUtils.zip(inDir, out);
     }
 
-    return createDescriptorFile(
+    DataSegment dataSegment = createDescriptorFile(
         segment.withLoadSpec(makeLoadSpec(outFile))
                .withSize(size)
                .withBinaryVersion(SegmentUtils.getVersionFromDir(inDir)),
         outFile.getParent(),
         fs
     );
+
+    Path outDir = new Path(String.format("%s/%s", config.getStorageDirectory(), storageDir));
+    fs.rename(outFile.getParent(), outDir);
```

Comment:
In case another task publishes the segment first, 
rename will return false and It seems we might be leaking the temporary directory above ?